tests/Doctrine/Tests/ORM/Functional/Ticket/GH7407Test.php

```diff
@@ -0,0 +1,95 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\Tests\ORM\Functional\Ticket;
+
+use Doctrine\ORM\UnitOfWork;
+use Doctrine\Tests\Models\CMS\CmsArticle;
+use Doctrine\Tests\Models\CMS\CmsUser;
+use Doctrine\Tests\OrmFunctionalTestCase;
+use ReflectionClass;
+
+use function spl_object_id;
+
+class GH7407Test extends OrmFunctionalTestCase
+{
+    protected function setUp(): void
+    {
+        $this->useModelSet('cms');
+
+        parent::setUp();
+    }
+
+    public function testMergingEntitiesDoesNotCreateUnmanagedProxyReferences(): void
+    {
+        // 1. Create an article with a user; persist, flush and clear the entity manager
+        $user           = new CmsUser();
+        $user->username = 'Test';
+        $user->name     = 'Test';
+        $this->_em->persist($user);
+
+        $article        = new CmsArticle();
+        $article->topic = 'Test';
+        $article->text  = 'Test';
+        $article->setAuthor($user);
+        $this->_em->persist($article);
+
+        $this->_em->flush();
+        $this->_em->clear();
+
+        // 2. Merge the user object back in:
+        // We get a new (different) entity object that represents the user instance
+        // which is now (through this object instance) managed by the EM/UoW
+        $mergedUser    = $this->_em->merge($user);
+        $mergedUserOid = spl_object_id($mergedUser);
+
+        // 3. Merge the article object back in,
+        // the returned entity object is the article instance as it is managed by the EM/UoW
+        $mergedArticle    = $this->_em->merge($article);
+        $mergedArticleOid = spl_object_id($mergedArticle);
+
+        // The $mergedArticle's #user property should hold the $mergedUser we obtained previously,
+        // since that's the only legetimate object instance representing the user from the UoW's
+        // point of view.
+        self::assertSame($mergedUser, $mergedArticle->user);
```

Comment:
I love this comment, and I think you should move it to the 3rd argument of that assertion, so that people can see it when it fails:

```suggestion
        self::assertSame($mergedUser, $mergedArticle->user, <<<'MESSAGE'
The $mergedArticle's #user property should hold the $mergedUser we obtained previously,
since that's the only legetimate object instance representing the user from the UoW's
point of view.
MESSAGE
);
```

There are probably other comments that would make good `$message` arguments below.

Reply:
I hate those heredocs for the visual breakage they cause, but I've changed comments to messages.